drivers/ethernet/eth_native_posix.c

```diff
@@ -231,55 +231,132 @@ static inline struct net_if *get_iface(struct eth_context *ctx,
 #endif
 }
 
-static int read_data(struct eth_context *ctx, int fd)
+#if defined(CONFIG_NET_VLAN)
+static struct net_pkt *prepare_vlan_pkt(struct eth_context *ctx,
+					int count, u16_t *vlan_tag, int *status)
 {
-	u16_t vlan_tag = NET_VLAN_TAG_UNSPEC;
-	struct net_if *iface;
+	struct net_eth_vlan_hdr *hdr = (struct net_eth_vlan_hdr *)ctx->recv;
 	struct net_pkt *pkt;
-	int count;
+#if defined(CONFIG_ETH_NATIVE_POSIX_VLAN_TAG_STRIP)
+	count -= NET_ETH_VLAN_HDR_SIZE;
+#endif
 
-	count = eth_read_data(fd, ctx->recv, sizeof(ctx->recv));
-	if (count <= 0) {
-		return 0;
+	pkt = net_pkt_rx_alloc_with_buffer(ctx->iface, count,
+					   AF_UNSPEC, 0, NET_BUF_TIMEOUT);
+	if (!pkt) {
+		*status = -ENOMEM;
+		return NULL;
 	}
 
+	net_pkt_set_vlan_tci(pkt, ntohs(hdr->vlan.tci));
+	*vlan_tag = net_pkt_vlan_tag(pkt);
+
+#if !defined(CONFIG_ETH_NATIVE_POSIX_VLAN_TAG_STRIP)
```

Comment:
why didn't you follow my comment there? You could avoid one net_pkt_write() in the code 

Also, I would advise to use a goto for error management: one place to unrek the packet, set the status and return NULL

Reply:
> why didn't you follow my comment there?

missed it, nothing else.

Reply:
> You could avoid one net_pkt_write() in the code

Couldn't follow this one.